bitcoin/src/crypto/sighash.rs

```diff
@@ -1318,6 +1208,183 @@ impl<E> EncodeSigningDataResult<E> {
+impl std::error::Error for TaprootError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        use TaprootError::*;
+
+        match *self {
+            PrevoutsSize(ref e) => Some(e),
+            PrevoutsIndex(ref e) => Some(e),
+            PrevoutsKind(ref e) => Some(e),
+            InputIndexOutOfBounds(ref e) => Some(e),
+            Io(_)
+            | SingleWithoutCorrespondingOutput { .. }
+            | InvalidSighashType(_) => None,
+        }
+    }
+}
+
+impl From<io::Error> for TaprootError {
+    fn from(e: io::Error) -> Self { Self::Io(e.kind()) }
+}
+
+impl From<transaction::IndexOutOfBoundsError> for TaprootError {
+    fn from(e: transaction::IndexOutOfBoundsError) -> Self { Self::InputIndexOutOfBounds(e) }
+}
+
+impl From<PrevoutsSizeError> for TaprootError {
+    fn from(e: PrevoutsSizeError) -> Self { Self::PrevoutsSize(e) }
+}
+
+impl From<PrevoutsKindError> for TaprootError {
+    fn from(e: PrevoutsKindError) -> Self { Self::PrevoutsKind(e) }
+}
+
+impl From<PrevoutsIndexError> for TaprootError {
+    fn from(e: PrevoutsIndexError) -> Self { Self::PrevoutsIndex(e) }
+}
+
+/// Annex must be at least one byte long and the first bytes must be `0x50`.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub struct AnnexError;
+
+impl fmt::Display for AnnexError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "annex must be at least one byte long and the first bytes must be `0x50`")
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for AnnexError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> { None }
+}
+
+/// Error computing a P2WPKH sighash.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum P2wpkhError {
+    /// Error computing the sighash.
+    Sighash(Error),
+    /// Script is not a witness program for a p2wpkh output.
+    NotP2wpkhScript,
+}
+
+impl fmt::Display for P2wpkhError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        use P2wpkhError::*;
+
+        match *self {
+            Sighash(ref e) => write_err!(f, "error encoding segwit v0 signing data"; e),
+            NotP2wpkhScript => write!(f, "script is not a script pubkey for a p2wpkh output"),
+        }
+    }
+}
+
+#[cfg(feature = "std")]
+impl std::error::Error for P2wpkhError {
+    fn source(&self) -> Option<&(dyn std::error::Error + 'static)> {
+        use P2wpkhError::*;
+
+        match *self {
+            Sighash(ref e) => Some(e),
+            NotP2wpkhScript => None,
+        }
+    }
+}
+
+impl From<Error> for P2wpkhError {
+    fn from(e: Error) -> Self { Self::Sighash(e) }
+}
+
+/// Error computing the sighash.
+#[derive(Debug, Clone, PartialEq, Eq)]
+#[non_exhaustive]
+pub enum Error {
+    /// Index out of bounds when accessing transaction input vector.
+    InputIndexOutOfBounds(transaction::IndexOutOfBoundsError),
+    /// Writer errored during consensus encoding.
+    Io(io::ErrorKind),
```

Comment:
Why is this `ErrorKind` and not `Error`?

Reply:
I wondered the same when I was doing this, I don't remember when this got changed/created.

Reply:
In a commit message for https://github.com/rust-bitcoin/rust-bitcoin/pull/2066:

>    Sadly the `std::io::Error` type requires `std::error::Error` as a
>    bound on the inner error, which is rather difficult to duplicate in
>    a way that allows for mapping to `std` and back.

Reply:
But why not our `io::Error`? We do have the type. I just tried it locally and it works if we drop the derives on errors.

Reply:
If it works we should switch to `io::Error` ... #2066 predates the `io` crate and is from an era when we were trying to glue core2 and std together in a transparent way, and it may be that we can do things now that we didn't use to be able to.

We use `io::ErrorKind` in a few places in this codebase. May be worth auditing those.